### PR TITLE
:hammer: A small accelerator refactor.

### PR DESCRIPTION
**Pull request by kartik4949**
## What does this PR do?
A small refactore in Accelerator_base class making it abstract class.

Fixes # (issue)
Asserting __setstate__ in accelerators.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by kartik4949** [comment]
@rohitgr7 @awaelchli ?

**Review by Borda — APPROVED** [review]
lgtm

**Comment by Borda** [comment]
@kartik4949 mind resolve conflicts and get tests to pass? :]

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -227,29 +241,30 @@ def get_reference_model(self, model) -> LightningModule:
...
-            'trainer': self.trainer,
-            'nickname': self.nickname,
-            'cluster_environment': self.cluster_environment,
-            'dist': self.dist,
-            'ddp_plugin': self.ddp_plugin
+            "trainer": self.trainer,
+            "nickname": self.nickname,
+            "cluster_environment": self.cluster_environment,
+            "dist": self.dist,
+            "ddp_plugin": self.ddp_plugin,
```
Comment (rohitgr7) [inline pytorch_lightning/accelerators/accelerator.py:248]: can you add `-S` to avoid string normalization here, in case you are using black?? These are unrelated changes to this PR.

Comment (kartik4949) [reply pytorch_lightning/accelerators/accelerator.py:248]: @rohitgr7 yes

File: pytorch_lightning/accelerators/accelerator.py [hunk 2]
```diff
@@ -33,6 +34,7 @@ class ReduceOp:
 
 
 class Accelerator(object):
+    __metaclass__ = ABCMeta
```
[the remark this answers is not in the document]
Comment (kartik4949) [reply pytorch_lightning/accelerators/accelerator.py:38]: yeah will subclass it with abc

**Comment by Borda** [comment]
@kartik4949 @justusschock how is it going here?

**Comment by kartik4949** [comment]
> @kartik4949 @justusschock how is it going here?
should i remove abstract? @Borda

**Comment by kartik4949** [comment]
> @kartik4949 @justusschock how is it going here?

@Borda @justusschock  removed abstract idea :) only setstate assertion

**Comment by kartik4949** [comment]
@Borda @justusschock only setstate assertion is left in this PR :)
we can merge i guess its ready

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -213,6 +213,7 @@ def __getstate__(self):
         }
 
     def __setstate__(self, d):
+        assert isinstance(d, dict), f"{type(d)} passed instead of dict."
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:216]: can you explain why this is needed? Lightning never calls this internal method, it's for pickling yes?
